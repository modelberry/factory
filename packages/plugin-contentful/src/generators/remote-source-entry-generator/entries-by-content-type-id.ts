import {
  ContentFields,
  ContentType,
  EntityMetaSysProps,
  SysLink,
} from 'contentful-management/types'

export type EntryTypeField = {
  contentFieldType?: ContentFields
  fieldValue: {
    /** Either syslink, an array of syslinks, a string Symbol value or a record of fieldIds */
    [localeCode: string]: SysLink | SysLink[] | Record<string, any> | string
  }
}

export type EntryTypeFields = {
  [fieldId: string]: EntryTypeField
}

export type EntryType = {
  fields: EntryTypeFields
  sys: EntityMetaSysProps
}

export type EntriesByContentTypeId = {
  [contentTypeId: string]: {
    contentType: ContentType
    entryTypeList: EntryType[]
  }
}
