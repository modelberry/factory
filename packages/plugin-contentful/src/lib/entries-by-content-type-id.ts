import {
  ContentFields,
  ContentType,
  SysLink,
} from 'contentful-management/types'

export type EntryTypeField = {
  contentFieldType?: ContentFields
  fieldValue: {
    /** Either syslink, an array of syslinks or a record of fieldIds */
    [localeCode: string]: SysLink | SysLink[] | Record<string, any>
  }
}

export type EntryTypeFields = {
  [fieldId: string]: EntryTypeField
}

export type Sys = {
  id: string
}

export type EntryType = {
  sys: Sys
  fields: EntryTypeFields
}

export type EntriesByContentTypeId = {
  [contentTypeId: string]: {
    contentType: ContentType
    entryTypeList: EntryType[]
  }
}
