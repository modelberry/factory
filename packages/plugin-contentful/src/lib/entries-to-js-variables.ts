import { EntryType } from '../pull-content/entries-by-content-type-id'
import { removeLinkAndLinkType } from '../pull-content/remove-link-and-link-type'

export interface EntriesToJsVariables {
  localeCode: string
  entryTypeList: EntryType[]
}

// Convert the Contentful entries into the a js content array
export const entriesToJsVariables = ({
  localeCode,
  entryTypeList,
}: EntriesToJsVariables) => {
  const contentArray: any[] = []
  // Add entries, convert arrays into Collections, etc.
  entryTypeList.forEach((entryType: EntryType) => {
    // Create a new entry with sys.id
    const addEntry: Record<string, any> = {
      sys: { id: entryType.sys.id },
    }
    // Add fields to entry
    Object.keys(entryType.fields).forEach((fieldId: string) => {
      // Remove type and linkType attributes because they can be resolved
      // by lookup up the content type
      const remoteEntryFieldValue = removeLinkAndLinkType({
        fieldValue: entryType.fields[fieldId].fieldValue[localeCode],
      })
      const remoteFieldType = entryType.fields[fieldId].contentFieldType
      if (remoteFieldType?.type === 'Array') {
        const itemsType =
          entryType.fields[fieldId].contentFieldType?.items?.type
        if (itemsType === 'Symbol') {
          // Same as the default below, but more explicit to code it this way
          addEntry[fieldId] = remoteEntryFieldValue
        } else {
          addEntry[`${fieldId}Collection`] = {
            items: remoteEntryFieldValue,
          }
        }
      } else if (remoteFieldType?.type === 'RichText') {
        // TODO: (maybe?) Convert all nodeTypes from string into numbers
        addEntry[fieldId] = {
          json: remoteEntryFieldValue,
        }
      } else {
        addEntry[fieldId] = remoteEntryFieldValue
      }
    })
    contentArray.push(addEntry)
  })
  return contentArray
}
