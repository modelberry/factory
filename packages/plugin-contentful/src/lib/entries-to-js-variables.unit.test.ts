import { EntryType } from './entries-by-content-type-id'
import { entriesToJsVariables } from './entries-to-js-variables'

const entryType: EntryType = {
  fields: {
    abstract: {
      contentFieldType: {
        id: 'abstract',
        localized: false,
        name: 'Abstract',
        required: false,
        type: 'Symbol',
      },
      fieldValue: {
        'en-US': 'This is the abstract',
      },
    },
    actions: {
      contentFieldType: {
        id: 'actions',
        items: {
          linkType: 'Entry',
          type: 'Link',
        },
        localized: false,
        name: 'Actions',
        required: false,
        type: '',
      },
      fieldValue: {
        'en-US': [
          {
            sys: {
              id: 'myContentActionId',
            },
          },
        ],
      },
    },
    heading: {
      contentFieldType: {
        id: 'heading',
        localized: false,
        name: 'Heading',
        required: false,
        type: 'Symbol',
      },
      fieldValue: {
        'en-US': 'Heres the heading',
      },
    },
  },
  sys: {
    contentType: {
      sys: {
        id: 'testTopic',
        linkType: 'ContentType',
        type: 'Link',
      },
    },
    createdAt: '2021-08-08T19:47:21.508Z',
    environment: {
      sys: {
        id: 'master',
        linkType: 'Environment',
        type: 'Link',
      },
    },
    id: '2v8GizBNh1oxbf6cKlpzJP',
    space: {
      sys: {
        id: 'tctsmza39jg6',
        linkType: 'Space',
        type: 'Link',
      },
    },
    type: 'Entry',
    updatedAt: '2021-08-08T19:47:21.878Z',
    version: 2,
  },
}

describe('entriesToJsVariables should', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  const entryTypeList: EntryType[] = [entryType]
  const localeCode = 'en-US'
  test('convert entries to js variables', async () => {
    const contentArray: any[] = entriesToJsVariables({
      entryTypeList,
      localeCode,
    })

    expect(contentArray).toMatchSnapshot()
  })
})
