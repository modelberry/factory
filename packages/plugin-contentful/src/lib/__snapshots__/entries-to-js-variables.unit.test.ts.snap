// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entriesToJsVariables should convert entries to js variables 1`] = `
Array [
  Object {
    "abstract": "This is the abstract",
    "actions": Array [
      Object {
        "sys": Object {
          "id": "myContentActionId",
        },
      },
    ],
    "heading": "Heres the heading",
    "sys": Object {
      "id": "2v8GizBNh1oxbf6cKlpzJP",
    },
  },
]
`;
